src/Bicep.Core/TypeSystem/Providers/ProviderSpecificationSyntaxFactory.cs

```diff
@@ -0,0 +1,82 @@
+
+    private const string BicepRegistryAddressPattern = @"br[:\/]\S+";
+
+    // Regex copied from https://semver.org/.
+    // language=regex
+    private const string SemanticVersionPattern = @"(0|[1-9]\d*)\.(0|[1-9]\d*)\.(0|[1-9]\d*)(?:-((?:0|[1-9]\d*|\d*[a-zA-Z-][0-9a-zA-Z-]*)(?:\.(?:0|[1-9]\d*|\d*[a-zA-Z-][0-9a-zA-Z-]*))*))?(?:\+([0-9a-zA-Z-]+(?:\.[0-9a-zA-Z-]+)*))?";
+
+    [GeneratedRegex(@$"^(?<name>{NamePattern})$", RegexOptions.ECMAScript | RegexOptions.Compiled)]
+    private static partial Regex ConfigManagedSpecificationPattern();
+
+    [GeneratedRegex(@$"^(?<name>{NamePattern})@(?<version>{SemanticVersionPattern})$", RegexOptions.ECMAScript | RegexOptions.Compiled)]
+    private static partial Regex LegacyBuiltInSpecificationPattern();
+
+    [GeneratedRegex(@$"^(?<address>{BicepRegistryAddressPattern})@(?<version>{SemanticVersionPattern})$", RegexOptions.ECMAScript | RegexOptions.Compiled)]
+    private static partial Regex InlinedSpecificationPattern();
+
+    [GeneratedRegex(@"^\S*[:\/](?<name>\S+)$", RegexOptions.ECMAScript | RegexOptions.Compiled)]
+    private static partial Regex RepositoryNamePattern();
+
+    public static IProviderSpecificationSyntax CreateProviderSpecificationSyntax(SyntaxBase syntax)
+     => syntax switch
+     {
+         StringSyntax stringSyntax when stringSyntax.TryGetLiteralValue() is { } value &&
+                                       TryCreateFromStringSyntax(stringSyntax, value) is { } specification
+             => specification,
+
+         IdentifierSyntax identifierSyntax
+             => new ConfigurationManagedProviderSpecificationSyntax(
+                 identifierSyntax.IdentifierName,
+                 IsValid: true,
+                 identifierSyntax.Span),
+
+         _ => new ProviderSpecificationSyntaxTrivia(syntax.Span)
+     };
+
+    private static IProviderSpecificationSyntax? TryCreateFromStringSyntax(StringSyntax stringSyntax, string value)
+    {
+        if (LegacyBuiltInSpecificationPattern().Match(value) is { Success: true } legacyMatch)
+        {
+            var name = legacyMatch.Groups["name"].Value;
+            var version = legacyMatch.Groups["version"].Value;
+            var span = new TextSpan(stringSyntax.Span.Position + 1, name.Length);
+
+            return new LegacyProviderSpecificationSyntax(name, version, IsValid: true, span);
+        }
+        else if (ConfigManagedSpecificationPattern().Match(value) is { Success: true } builtInMatch)
+        {
+            var name = builtInMatch.Groups["name"].Value;
+            var span = new TextSpan(stringSyntax.Span.Position + 1, name.Length);
+
+            return new ConfigurationManagedProviderSpecificationSyntax(name, IsValid: true, span);
+        }
+        else if (InlinedSpecificationPattern().Match(value) is { Success: true } registryMatch)
+        {
+            // NOTE(asilverman): The regex for the registry pattern is intentionally loose since it will be validated by the module resolver.
```

Comment:
[nit] We've been keeping internal aliases out of comments - reword to just:
```csharp
// The regex for the registry pattern is intentionally loose since it will be validated by the module resolver.
```

Reply:
asilverman is my public github alias, but I can remove...

Reply:
Thanks. Yeah I think as a general principal, I'd rather keep comments anonymous. If someone wants to figure out who wrote the comment, they can always use `git blame`.